File: csharp/1010.pairs-of-songs-with-total-durations-divisible-by-60.cs
/*
 * @lc app=leetcode id=1010 lang=csharp
 *
 * [1010] Pairs of Songs With Total Durations Divisible by 60
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int NumPairsDivisibleBy60(int[] time) {
        var mins = time
            .GroupBy(t => t % 60, (int t, IEnumerable<int> all) => (Time: t, Count: all.Count()))
            .ToDictionary(tuple => tuple.Time, tuple => tuple.Count);
        return mins
            .Where(p => p.Key <= 30)
            .Select(p => p.Key == 0 || p.Key == 30 ? p.Value * (p.Value - 1) / 2 : (mins.TryGetValue(60 - p.Key, out int o) ? o * p.Value : 0))
            .Sum();
    }
}
// @lc code=end

File: java/1010.pairs-of-songs-with-total-durations-divisible-by-60.java
/*
 * @lc app=leetcode id=1010 lang=java
 *
 * [1010] Pairs of Songs With Total Durations Divisible by 60
 */

// @lc code=start
class Solution {
    public int numPairsDivisibleBy60(int[] time) {
        int[] count = new int[60];
        int ret = 0;
        for (int t : time) {
            int i = t % 60;
            ret += count[(60 - i) % 60];
            count[i]++;
        }
        return ret;
    }
}
// @lc code=end

